_posts/2023-08-25-using-kani-to-validate-security-boundaries-in-aws-firecracker.md

```diff
@@ -0,0 +1,292 @@
+
+However, when trying to stub `now`, one will quickly notice that `Instant` does not offer any constructors for creating an instance from, say, a Unix timestamp. In fact, it is impossible to construct an `Instant` outside of the standard library as its fields are private. When in such a situation, the solution is often to go down the call stack of the function that you want to stub, to see if any of the functions further down can be stubbed out instead to achieve the desired effect. In our case, `now` calls functions in (private) OS specific time modules, until it bottoms out at [`libc::clock_gettime`](https://www.gnu.org/software/libc/manual/html_node/Getting-the-Time.html#index-clock_005fgettime).
+
+The `clock_gettime` function is passed a pointer to a `libc::timespec` structure, and the `tv_sec` and `tv_nsec` members of this structure are later used to construct the `Instant` returned by `Instant::now`. Therefore, we can use the following stub to achieve our goal of getting non-deterministic, monotonically non-decreasing `Instant`s:
+
+```rs
+mod stubs {
+    static mut LAST_SECONDS: i64 = 0;
+    static mut LAST_NANOS: i64 = 0;
+        
+    const NANOS_PER_SECOND: i64 = 1_000_000_000;
+
+    pub unsafe extern "C" fn clock_gettime(_clock_id: libc::clockid_t, tp: *mut libc::timespec) -> libc::c_int {
+        // kani::any_where provides us with a non-deterministic number of seconds
+        // that is at least equal to LAST_SECONDS (to ensure that time only
+        // progresses forward).
+        let next_seconds = kani::any_where(|&n| n >= unsafe { LAST_SECONDS });
+        let next_nanos = kani::any_where(|&n| n >= 0 && n < NANOS_PER_SECOND);
+
+        if next_seconds == unsafe { LAST_SECONDS } {
+            kani::assume(next_nanos >= unsafe { LAST_NANOS });
+        }
+
+        unsafe {
+            (*tp).tv_sec = LAST_SECONDS;
+            (*tp).tv_nsec = LAST_NANOS;
+        }
+
+        unsafe {
+            LAST_SECONDS = next_seconds;
+            LAST_NANOS = next_nanos;
+        }
+
+        0
+    }
+}
+```
+
+Note how the first invocation of this stub will always set `tv_sec = tv_nsec = 0`, as this is what the statics are initialized to. This is an optimization we do because the rate-limiter only cares about the delta between two instants, which will be non-deterministic as long as one of the two instants is non-deterministic. In order to keep Kani performant, it is important to minimize the number of non-deterministic values, especially if multiplication and division is involved.
```

Comment:
```suggestion
Note how the first invocation of this stub will always set `tv_sec = tv_nsec = 0`, as this is what the statics are initialized to. This is an optimization we do because the rate limiter only cares about the delta between two instants, which will be non-deterministic as long as one of the two instants is non-deterministic. **In order to keep Kani performant, it is important to minimize the number of non-deterministic values, especially if multiplication and division is involved.**
```